### ISSUE TITLE
Send offline -> POST /transaction/send-offline

### ISSUE THREAD

**Issue by michaelredman** [paritytech/substrate-api-sidecar#419]
I would be great to see "POST /transaction/send-offline"

GET /transaction/material is useful, however a better endpoint would be

```
POST /transaction/payload

{
   "from"...
   "to"...
   "amount"...
   "method"..
}
```

The method could do exactly what polkadot-js-signer does for sendOffline e.g

```
polkadot-js-signer sendOffline --account xxx --ws ws://localhost:9944/ balances.transfer yyyyy 63333 
```

The response JSON wuld be

```
{
"payload": \"0x33333333333333"
}
```

I can sign the payload and then submit to the existing POST /transaction. This means I do not have to go through the ordeal of having to create the extrinsic myself.

**Comment by emostov** [comment]
Thoughts @joepetrowski ?

**Comment by michaelredman** [comment]
I have made this change, following the workflow of polkadot-js-signer, I can raise a PR  if you think it would be useful. Hiding the protocol details makes it much more useable, you only need to know how to sign a hex string and do a HTTP POST.

**Comment by joepetrowski** [comment]
Wow that was fast! We've been discussing the last day and are definitely interested in this change. I think I would use the name `POST transaction/constructUnsigned` or `constructPayload` since it's not directly used for sending, but rather signing. Naming aside though, I think this would be a great feature.

**Comment by emostov** [comment]
My main concern is just making sure users can parse the transactions before signing them, which requires them to have type definitions/ metadata etc on their end. Not being able to/ not trying to parse transactions before signing can result in some heinous attacks (e.g. you unwittingly sign and broadcast a tx that sends all you funds to an attacker).

**Comment by michaelredman** [comment]
So the workflow I have here in the following PR (https://github.com/paritytech/substrate-api-sidecar/pull/420) has no parsing.

I provide the JSON metada used for the signature, you just need to know how to sign the hex string in an algo. No protocol information is needed.

There is one method I have left for someone with more substrate knowledge than me. It is in the file types/Signature.ts

I am unable to properly decode the metadata used for signing (which means you need to be fast... if a block rolls past).

```
export function deserialize(signerOptions: SignatureOptions): Partial<SignerOptions> {
    return {
        blockHash: signerOptions.blockHash,
        nonce: signerOptions.nonce,
        era: signerOptions.era,
        tip: 0
    }
}

export function serialize(signerOptions: Partial<SignerOptions>): SignatureOptions {
    return {
        blockHash: signerOptions.blockHash,
        nonce: signerOptions.nonce,
        era: signerOptions.era,
        tip: 0
    }
}
```

**Comment by michaelredman** [comment]
Happy to change the path from /transactions/send-offline, to /transactions/construct or /transactions/prepare or whatever.

**Comment by michaelredman** [comment]
Would appeciate tips on how to implement

```
export function deserialize(signerOptions: SignatureOptions): Partial<SignerOptions> {
    return {
        blockHash: signerOptions.blockHash,
        nonce: signerOptions.nonce,
        era: signerOptions.era,
        tip: 0
    }
}

export function serialize(signerOptions: Partial<SignerOptions>): SignatureOptions {
    return {
        blockHash: signerOptions.blockHash,
        nonce: signerOptions.nonce,
        era: signerOptions.era,
        tip: 0
    }
}
```

I want to get on with some work, and having these methods not working correctly means I need to sign really quickly, which is a PITA.

**Comment by emostov** [comment]
Thank you for putting in all the time and the work so far. 

You mention
> you just need to know how to sign the hex string

My concern is that users who think this and not bother decoding the exact payload they sign will be sitting ducks for a man in the middle attack.

To elaborate on my concern about decoding I will outline a sample man in the middle attack attack:

1) I send a POST request to `/transaction/send-offline` for a balances.transferKeepAlive from me to Alice.
2) I get a response from Sidecar where the `unsigned` field is a hex encoded string for a transfer from me to an attacker (presumably the attacker intercepted the request and put there own tx in)
```JSON
{
  "phase": "SignatureRequired",
  "payload": {
    "unsigned": "0x<encoded balance transfer to hacker>",
    "options": {
        ...
    }
  }
}
```
3) I sign the payload and respond with 
```
{
  "account": "5F6h1SnxKrNANfuGdUrYwht8VXoGdxW3xVXPD6j1mxXbVdgj",
  "target": "balances.transfer",
  "params": [
    "<Alice's address>",
    "01234"
  ],
  "signature": {
    "type": "sr25519",
    "value": "0x0628be214d545df7cdc2021371fb4eea8d7317411f61c676b437d9fbd805eb1adc842c02c29bc3e08ef4ee3e23012a1ae63dc76666624538ba3eb97d13ec5589",
    "options": {
       ...
     },
  }
}
```
However, the **attacker intercepts my response and stuffs it with their address instead of Alices**, and since I did not verify the payload I signed I could not tell it had the attackers address.

**Comment by michaelredman** [comment]
You could return a signature in the payload which is blake(unsigned, from, to, balance) to verify.

**Comment by michaelredman** [comment]
There are plenty of ways to provide verification here, the idea is not to expose the changing schemas and protocol. A gateway should provide an abstraction that is easy to use. If you have suitable encoders/decoders you do not need the API sidecar.

**Comment by emostov** [comment]
Thank you for the thoughtful responses. Still some more follow up questions to make sure I understand.

In this verification scheme what is keeping the attacker from creating
```
    "unsigned": "0x<encoded balance transfer to hacker>",
    "hash": {
        "value": "HASH("0x<encoded balance transfer to hacker>, from, to, balance)",
        "type": "blake" | "sha256"|...
    } 
```

So when I verified with
```
if(response.payload.hash != blake(response.payload.unsigned, request.from, request,to, request.amount)) {
   alert("The hash does not match, how could anyone compose a blake hash from my request to match");
}
```
it would still pass because `response.payload.unsigned` is the same everywhere?

From what I can tell there is nothing keeping the attacker from hashing there malicious `unsigned` + my legit params?

**Comment by michaelredman** [comment]
If you use an PKI hash, like RSA for example. So if the public key is available, and the sidecar does a HASH with the private key, then you know the man in the middle can't create the hash? All you would need is a PEM text file on the sidecar.

**Comment by michaelredman** [comment]
Better yet, encrypt it with the public key of the account. If you can sign it you can decrypt it.

Polkadot addresses/accounts are public keys can you just encrypt the unsigned in the public key? Only the owner of the private key can decrypt it.

All of these schemes are possible, I don't see an issue with the workflow.

**Comment by michaelredman** [comment]
Another option, HMAC + SHA256 and a shared secret, useful for private deployments. You can setup the service with a config option for a simple string secret. 

Behind HTTPS where you trust the site, identity hash, which is basically nothing. With usernames and password, salt the HMAC + SHA256 with a hash of the password.. etc.. etc..

```
    "unsigned": "0x<encoded balance transfer to hacker>",
    "hash": {
        "value": "HASH("0x<encoded balance transfer to hacker>, from, to, balance)",
        "type": "rsa" | "hmacsha256" | "none"
    } 
```

**Comment by michaelredman** [comment]
Any scheme found here could be used https://jwt.io/ a JWT basically offers the same thing.

```
hashing-scheme.info.hash
```

In order to trust ```info``` you regenerate it with ```hashing-scheme``` if its equal you can trust it. Same thing with the above.

**Comment by emostov** [comment]
The functionality you have proposed so far are sound and I think a service like this would provide significant utility to the substrate ecosystem. We would support the further development of the concept. However, as  one of the maintainers I currently have 2 blockers to going forward with this work right now in api-sidecar

1) Should a more complex, stateful functionality like this live in api-sidecar, which is right now designed to be lightweight, stateless and easy to setup? Or should a new service be created (e.g. substrate-tx-gateway) that focuses on supporting endpoints that would require something akin to jwt and be designed exclusively for more advanced users?
2) While Parity may be able to dedicate resources to this in the future (and would like to), we have some higher priority projects and it could be up to a few months until we are able to give this proper attention

If you want to go forward with it here on your own time feel free; I can probably help out here and there with the encoding and decoding bits. But even with a working branch I can’t guarantee we would merge it in here. (There are certainly numerous options if we don’t want the functionally directly in api-sidecar, like creating another microservice that lives in the repo etc., but this requires more discussion to see what makes sense)

**Comment by michaelredman** [comment]
Sure, one thing that would be useful, is how to serialize and deserialize the substrate types e.g

```
export function deserialize(signerOptions: SignatureOptions): Partial<SignerOptions> {
    return {
        blockHash: signerOptions.blockHash,
        nonce: signerOptions.nonce,
        era: signerOptions.era,
        tip: 0
    }
}

export function serialize(signerOptions: Partial<SignerOptions>): SignatureOptions {
    return {
        blockHash: signerOptions.blockHash,
        nonce: signerOptions.nonce,
        era: signerOptions.era,
        tip: 0
    }
}
```

Is there an easy way to do this, for instance if I create an Era or BlockHash via createType, how do I represent as JSON and then recover the identical type from that same JSON string. 

I am assuming there is some utility for this?

**Comment by typtyptyp2007** [comment]
> I would be great to see "POST /transaction/send-offline"
> 
> GET /transaction/material is useful, however a better endpoint would be
> 
> ```
> POST /transaction/payload
> 
> {
>    "from"...
>    "to"...
>    "amount"...
>    "method"..
> }
> ```
> 
> The method could do exactly what polkadot-js-signer does for sendOffline e.g
> 
> ```
> polkadot-js-signer sendOffline --account xxx --ws ws://localhost:9944/ balances.transfer yyyyy 63333 
> ```
> 
> The response JSON wuld be
> 
> ```
> {
> "payload": \"0x33333333333333"
> }
> ```
> 
> I can sign the payload and then submit to the existing POST /transaction. This means I do not have to go through the ordeal of having to create the extrinsic myself.

Hello,Where can I find this interface?
I need it very much.

**Comment by emostov** [comment]
Not sure if this is still an issue but here are some starting points: 

**tl;dr** you want to look at `api.createType(<RuntimeType>, <scale encoded | polkadot-js types | json>)`

```
export function deserialize(signerOptions: SignatureOptions): Partial<SignerOptions> {
    return {
        blockHash: signerOptions.blockHash,
        nonce: signerOptions.nonce,
        era: signerOptions.era,
        tip: 0
    }
}
```

I am not sure what you mean exactly by signature options, but for a encoded tx payload at any stage you can do something like the decode methods in txwrapper: https://github.com/paritytech/txwrapper/blob/master/src/decode/decode.ts

```
export function serialize(signerOptions: Partial<SignerOptions>): SignatureOptions {
    return {
        blockHash: signerOptions.blockHash,
        nonce: signerOptions.nonce,
        era: signerOptions.era,
        tip: 0
    }
}
```

Again not exactly sure what you are trying to accomplish here, but if you want to serialize a signing payload with the signature to create a ready-to-submit Extrinsic (see: `UncheckedExtrinsic` in substrate source) you can do something like this: 
https://github.com/paritytech/txwrapper/blob/c7b18f7289f68fdd90b4715339b6f0629aa315e9/src/createSignedTx.ts#L21-L27

**Comment by emostov** [comment]
> > I would be great to see "POST /transaction/send-offline"
> > GET /transaction/material is useful, however a better endpoint would be
> > ```
> > POST /transaction/payload
> > 
> > {
> >    "from"...
> >    "to"...
> >    "amount"...
> >    "method"..
> > }
> > ```
> > 
> > 
> > The method could do exactly what polkadot-js-signer does for sendOffline e.g
> > ```
> > polkadot-js-signer sendOffline --account xxx --ws ws://localhost:9944/ balances.transfer yyyyy 63333 
> > ```
> > 
> > 
> > The response JSON wuld be
> > ```
> > {
> > "payload": \"0x33333333333333"
> > }
> > ```
> > 
> > 
> > I can sign the payload and then submit to the existing POST /transaction. This means I do not have to go through the ordeal of having to create the extrinsic myself.
> 
> Hello,Where can I find this interface?
> I need it very much.

As [mentioned here](https://github.com/paritytech/substrate-api-sidecar/issues/419#issuecomment-774238101) we have chosen not to support this feature in substrate-api-sidecar for time being. I am not personally aware of any service that does this yet, but others in the community may be working on something that does this.

**Comment by TarikGul** [comment]
Closing.
